### DISCUSSION TITLE
Why the distinction between mutations and queries?

### DISCUSSION THREAD

**Discussion by sbdchd** [trpc/trpc#1638] [Q&A] [answered]
The docs (https://trpc.io/docs/quickstart#add-a-mutation-endpoint) mention:

> Similarly to GraphQL, tRPC makes a distinction between query and mutation endpoints.

I was wondering what the motivation for this is, easier middlewares?

**Comment by KATT** [comment, marked as answer]
Main reasons:

- semantics -- it's a lot easier to reason about your codebase with the distinction
- queries happens through HTTP `GET` --> easy HTTP caching
- mutations happens through HTTP `POST` --> not cached

**Reply by boredland** [reply]
Follow-Up question: I currently have a pretty giant list of query parameters (>200 hex codes, don't ask), that sometimes is so huge, it seems to exceed browser limits. Would it be possible to send queries using POST in such edge-cases?

**Reply by jlalmes** [reply]
That would not be possible, but surely in this case you'd be better off just changing your `query` procedure into a `procedure` so you don't hit the URL limits?

**Reply by boredland** [reply]
you mean like [this](https://trpc.io/docs/useInfiniteQuery#example-procedure) using infinitequery?

**Reply by jlalmes** [reply]
I don't know your exact use-case. But you likely have 2 options.

1. Make multiple requests to your `query` procedure with your hex codes in smaller batches.
2. Change your `query` into a `mutation` procedure so that the hex codes are passed in the request body & you will never hit the URL limits.

I would choose option 2. I can't imagine any tradeoff changing your request from `GET` -> `POST`.

**Reply by KATT** [reply]
Threw this out on Twitter now :) https://twitter.com/alexdotjs/status/1528854148011130880

**Reply by boredland** [reply]
Just to document this for future readers: @sachinraja proposed to wrap a mutation using the [vanilla client.mutation](https://trpc.io/docs/vanilla) inside a regular [useQuery-Hook](https://react-query.tanstack.com/reference/useQuery) and I think this is the easiest possible workaround for now!

**Reply by KATT** [reply]
Yeah, I do that for some complex multi-step mutations as well :)

**Reply by boredland** [reply]
is it easily possible to re-use your great loggerlink setup with the vanilla useQuery?

**Reply by KATT** [reply]
It should be there automatically!

**Reply by KATT** [reply]
Try this (I haven't tested it)

```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { TRPCClientErrorLike } from '@trpc/client';
import { UseMutationOptions, useMutation } from 'react-query';
import { AppRouter } from '~/server/routers/_app';
import { inferQueryInput, inferQueryOutput, trpc } from './trpc';

export function useQueryAsMutation<
  TPath extends keyof AppRouter['_def']['queries'],
>(
  path: TPath,
  opts?: UseMutationOptions<
    inferQueryOutput<TPath>,
    TRPCClientErrorLike<AppRouter>,
    inferQueryInput<TPath>
  >,
) {
  const utils = trpc.useContext();

  return useMutation<
    inferQueryOutput<TPath>,
    TRPCClientErrorLike<AppRouter>,
    inferQueryInput<TPath>
  >(`__${path}`, (input) => utils.fetchQuery([path, input] as any), opts);
}


```
